system/src/control/storage.cpp

```diff
@@ -191,74 +173,24 @@ int firmwareUpdateDataRequest(ctrl_request* req) {
-            }
-            const auto type = moduleFunctionToPb((module_function_t)module.info.module_function);
-            if (type == PB(FirmwareModuleType_INVALID_FIRMWARE_MODULE)) {
-                continue;
-            }
-            PB(GetModuleInfoReply_Module) pbModule = {};
-            pbModule.type = type;
-            pbModule.index = module.info.module_index;
-            pbModule.version = module.info.module_version;
-            pbModule.size = (uintptr_t)module.info.module_end_address - (uintptr_t)module.info.module_start_address;
-            unsigned valid = 0;
-            if (!(module.validity_result & MODULE_VALIDATION_INTEGRITY)) {
-                valid |= PB(FirmwareModuleValidityFlag_INTEGRITY_CHECK_FAILED);
-            }
-            if (!(module.validity_result & MODULE_VALIDATION_DEPENDENCIES)) {
-                valid |= PB(FirmwareModuleValidityFlag_DEPENDENCY_CHECK_FAILED);
-            }
-            pbModule.validity = valid;
-            pbModule.dependencies.arg = const_cast<module_info_t*>(&module.info);
-            pbModule.dependencies.funcs.encode = [](pb_ostream_t* strm, const pb_field_iter_t* field, void* const* arg) {
-                const auto info = (const module_info_t*)*arg;
-                for (unsigned i = 0; i < 2; ++i) {
-                    const module_dependency_t& dep = (i == 0) ? info->dependency : info->dependency2;
-                    const auto type = moduleFunctionToPb((module_function_t)dep.module_function);
-                    if (type == PB(FirmwareModuleType_INVALID_FIRMWARE_MODULE)) {
-                        continue;
-                    }
-                    PB(GetModuleInfoReply_Dependency) pbDep = {};
-                    pbDep.type = type;
-                    pbDep.index = dep.module_index;
-                    pbDep.version = dep.module_version;
-                    if (!pb_encode_tag_for_field(strm, field)) {
-                        return false;
-                    }
-                    if (!pb_encode_submessage(strm, PB(GetModuleInfoReply_Dependency_fields), &pbDep)) {
-                        return false;
-                    }
-                }
-                return true;
-            };
-            if (!pb_encode_tag_for_field(strm, field)) {
-                return false;
-            }
-            if (!pb_encode_submessage(strm, PB(GetModuleInfoReply_Module_fields), &pbModule)) {
-                return false;
-            }
-        }
-        return true;
-    };
+    using namespace particle::system;
+    EncodeFirmwareModules modules(&pbRep.modules, EncodeFirmwareModules::Flag::FORCE_SHA);
     CHECK(encodeReplyMessage(req, PB(GetModuleInfoReply_fields), &pbRep));
     return 0;
 }
 
+#if HAL_PLATFORM_ASSETS
+int getAssetInfo(ctrl_request* req) {
+    PB(GetAssetInfoReply) pbRep = {};
+    using namespace particle::system;
+    EncodeAssets assetsAvailable(&pbRep.available, AssetManager::instance().availableAssets());
```

Comment:
Similarly I think this should use `availableAndRequiredAssets()` for consistency with the describe and Wiring API. The CLI flashes the new application before assets so returning available and required assets in the control request will work.

Reply:
I'd disagree. I'd like for system describe and control requests to represent the full picture and for wiring APIs to be representative of what's required by the application. Again, same thing: to avoid re-flashing things unnecessarily that are already on the device.